### PR TITLE
Cluster Registry

### PR DESCRIPTION
**Pull request by JumaX**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

This PR adds support for Cluster Registry across Brokers, Schema Registry, KSQL, Kafka Connect as per documentation here: https://docs.confluent.io/current/security/cluster-registry.html

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Have validated against rbac-mtls-rhel scenario and added another check in the verify.yml to validate that the cluster ID's registered in the inventory file match what is in the cluster registry.

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

### CODE REVIEW

File: roles/confluent.test/tasks/check_cluster_registry.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    status_code: 200
+  register: cluster_name_query
+
+- set_fact:
+    cluster_name_query_json: "{{ cluster_name_query.json }}"
+
+- name: Assert Broker Cluster ID as expected
+  assert:
+    that:
+      - broker_id == "{{cluster_name_query_json | json_query('[0].clusterName')}}"
```
Comment (domenicbove) [inline roles/confluent.test/tasks/check_cluster_registry.yml:24]: Great work with this PR! Only minor thing is doesn't this filter have a python dependency on jmes-path:
https://docs.ansible.com/ansible/latest/user_guide/playbooks_filters.html#selecting-json-data-json-queries

Is there another way to pull out each element? maybe 
```
      - broker_id == "{{cluster_name_query_json[0].clusterName}}"
```

what does this return:
```
- debug: var= cluster_name_query_json
```

File: roles/confluent.test/tasks/check_cluster_registry.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    fail_msg: "Connect Cluster ID set to {{cluster_name_query_json | json_query('[1].clusterName')}} not {{connect_id}}"
+    quiet: true
+
+- name: Assert KSQL Cluster ID as expected
+  assert:
+    that:
+      - ksql_id == "{{cluster_name_query_json | json_query('[2].clusterName')}}"
+    fail_msg: "KSQL Cluster ID set to {{cluster_name_query_json | json_query('[2].clusterName')}} not {{ksql_id}}"
+    quiet: true
+
```
Comment (domenicbove) [inline roles/confluent.test/tasks/check_cluster_registry.yml:41]: no Rest Proxy cluster registry?

File: roles/confluent.schema_registry/tasks/main.yml [hunk 3]
```diff
@@ -191,3 +191,7 @@
   when:
     - health_checks_enabled|bool
     - not ansible_check_mode
+
+- name: Register Cluster
+  include_tasks: register_cluster.yml
+  when: cluster_registry_cluster_name is defined
```
Comment (domenicbove) [inline roles/confluent.schema_registry/tasks/main.yml:197]: can we name this variable something tied to the component like `schema_registry_cluster_registry_name`
open to suggestions

also same goes for the other components

File: roles/confluent.test/tasks/check_cluster_registry.yml (hunk 1 above)
Comment (JumaX) [reply roles/confluent.test/tasks/check_cluster_registry.yml:24]: I tried this and it didn't work, don't see why the dependency is an issue.

File: roles/confluent.test/tasks/check_cluster_registry.yml (hunk 2 above)
Comment (JumaX) [reply roles/confluent.test/tasks/check_cluster_registry.yml:41]: Good question, we don't mention it in the cluster registry documentation at all, and Rest Proxy itself does not support clustering, in that each instance is unique and does not/can not rely on any other instance.  https://docs.confluent.io/current/kafka-rest/production-deployment/rest-proxy/index.html#deployment

That being said, you still may want to grant each instance a unique/simple name.  Let me ask the team.

File: roles/confluent.schema_registry/tasks/main.yml (hunk 3 above)
Comment (JumaX) [reply roles/confluent.schema_registry/tasks/main.yml:197]: Happy to change to component_name_cluster_name if that suits?  So for example schema_registry_cluster_name

Comment (domenicbove) [reply roles/confluent.schema_registry/tasks/main.yml:197]: ya! I think `component_name_cluster_name` works for me. We will want them to have the component name in there bc what if there is service colocation

oh also these variables need documentation, which may be kinda challenging. idk what happens if you have something like this in `confluent.variables/defaults/main.yml`:
```
### Cluster registry cluster name for Schema Registry cluster
schema_registry_cluster_name: 
```

If that will mean schema_registry_cluster_name is defined or not. gotta test.

Worst case you could have:
```
schema_registry_cluster_name: ""
```

and then have the conditional be:
```
when: schema_registry_cluster_name|length >0
```

But i like the defined option more

File: roles/confluent.test/tasks/check_cluster_registry.yml (hunk 1 above)
Comment (domenicbove) [reply roles/confluent.test/tasks/check_cluster_registry.yml:24]: what does this return:
```
- debug: var=cluster_name_query_json
```

Comment (JumaX) [reply roles/confluent.test/tasks/check_cluster_registry.yml:24]: I've already done that.  The first thing I did was dump the contents of the var to get the structure of the json.  I then tried selecting individual elements but it was really painful to get it to work.  So then I tried the json query is it's much simpler and cleaner.

Unless there is a specific technical reason why json query is not valid here, I think it should stay as is.

File: roles/confluent.test/tasks/check_cluster_registry.yml (hunk 2 above)
Comment (JumaX) [reply roles/confluent.test/tasks/check_cluster_registry.yml:41]: So confirmed, we don't model anything in cluster registry for Rest Proxy as it also doesn't have dedicated RBAC requirements either.

File: roles/confluent.test/tasks/check_cluster_registry.yml (hunk 1 above)
Comment (domenicbove) [reply roles/confluent.test/tasks/check_cluster_registry.yml:24]: The reason not to add dependencies to our list is we will have to update the how to test docs, and potentially the jenkins image as well. 

I'd like to see the output of this:
```
- debug: var=cluster_name_query_json
```
so i can try and help!

Comment (domenicbove) [reply roles/confluent.test/tasks/check_cluster_registry.yml:24]: this works: `cluster_name_query_json[0]['clusterName']`

File: roles/confluent.variables/defaults/main.yml [hunk 4]
```diff
@@ -355,6 +355,9 @@ kafka_broker_custom_properties: "{{ kafka_broker.properties }}"
 ### Boolean to enable the embedded rest proxy within Kafka. NOTE- Embedded Rest Proxy must be enabled if RBAC is enabled and Confluent Server must be enabled
 kafka_broker_rest_proxy_enabled: "{{confluent_server_enabled}}"
 
+### Use to register and identify your Kafka clusters in the MDS using a name of your choosing.
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:358]:
```suggestion
### Set to the name of your Kafka cluster to be registered in the MDS. If you do not set this variable, cluster will not be registered.
```

Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:358]:
```suggestion
### Use to register and identify your Kafka cluster in the MDS.
```

File: roles/confluent.variables/defaults/main.yml [hunk 5]
```diff
@@ -425,6 +428,8 @@ schema_registry_copy_files: []
 ### Use to set custom schema registry properties. This variable is a dictionary. Put values true/false in quotation marks to perserve case. NOTE- kafka_broker.properties is deprecated.
 schema_registry_custom_properties: "{{ schema_registry.properties }}"
 
+### Use to register and identify your Schema Registry clusters in the MDS using a name of your choosing.
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:431]:
```suggestion
### Use to register and identify your Schema Registry cluster.
```

File: roles/confluent.variables/defaults/main.yml [hunk 6]
```diff
@@ -592,6 +597,8 @@ kafka_connect_custom_properties: "{{ kafka_connect.properties }}"
 ### Boolean to configure Monitoring Interceptors on Connect.
 kafka_connect_monitoring_interceptors_enabled: "{{ monitoring_interceptors_enabled }}"
 
+### Use to register and identify your Kafka Connect cluster in the MDS using a name of your choosing.
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:600]:
```suggestion
### Use to register and identify your Kafka Connect cluster in the MDS.
```

File: roles/confluent.variables/defaults/main.yml [hunk 7]
```diff
@@ -671,6 +678,8 @@ ksql_custom_properties: "{{ ksql.properties }}"
 ### Boolean to configure Monitoring Interceptors on ksqlDB.
 ksql_monitoring_interceptors_enabled: "{{ monitoring_interceptors_enabled }}"
 
+### Use to register and identify your KSQL clusters in the MDS using a name of your choosing.
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:681]:
```suggestion
### Use to register and identify your KSQL clusters in the MDS.
```

File: roles/confluent.variables/defaults/main.yml (hunk 5 above)
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:431]:
```suggestion
### Use to register and identify your Schema Registry cluster in the MDS.
```

File: roles/confluent.test/molecule/rbac-mtls-rhel/verify.yml [hunk 8]
```diff
@@ -267,3 +267,20 @@
         service_ca: /var/ssl/private/ca.crt
         service_cert: /var/ssl/private/kafka_connect.crt
         service_key: /var/ssl/private/kafka_connect.key
+
+- name: Validate Cluster Registry
+  hosts: schema_registry
+  gather_facts: false
+  tasks:
```
Comment (domenicbove) [inline roles/confluent.test/molecule/rbac-mtls-rhel/verify.yml:274]: please move the check_cluster_registry.yml tasks here, and remove that file. we can make a generic file later if need be!

File: roles/confluent.variables/defaults/main.yml [hunk 9]
```diff
@@ -671,6 +678,8 @@ ksql_custom_properties: "{{ ksql.properties }}"
 ### Boolean to configure Monitoring Interceptors on ksqlDB.
 ksql_monitoring_interceptors_enabled: "{{ monitoring_interceptors_enabled }}"
 
+### Use to register and identify your KSQL clusters in the MDS.
```
Comment (domenicbove) [inline roles/confluent.variables/defaults/main.yml:681]:
```suggestion
### Use to register and identify your KSQL cluster in the MDS.
```
```suggestion
### Use to register and identify your KSQL clusters in the MDS.
```

**Review by domenicbove — APPROVED** [review]

**Comment by domenicbove** [comment]
Nice work on this one!

**Pull request merged**